lib/core/test/unit/Cardano/Wallet/Primitive/Migration/PlanningSpec.hs

```diff
@@ -0,0 +1,418 @@
+    , shrinkMockInput
+    , unMockTxConstraints
+    )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle (..) )
+import Control.Monad
+    ( replicateM )
+import Data.Either
+    ( isLeft, isRight )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.Set
+    ( Set )
+import Fmt
+    ( padLeftF, pretty )
+import GHC.Generics
+    ( Generic )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.Hspec.Core.QuickCheck
+    ( modifyMaxSuccess )
+import Test.Hspec.Extra
+    ( parallel )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Blind (..)
+    , Gen
+    , Property
+    , checkCoverage
+    , choose
+    , counterexample
+    , cover
+    , oneof
+    , property
+    , shrinkList
+    , tabulate
+    , withMaxSuccess
+    , (===)
+    )
+
+import qualified Cardano.Wallet.Primitive.Migration.Selection as Selection
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as Set
+
+spec :: Spec
+spec = describe "Cardano.Wallet.Primitive.Migration.PlanningSpec" $
+
+    modifyMaxSuccess (const 1000) $ do
+
+    parallel $ describe "Creating migration plans" $ do
+
+        describe "Small migrations" $
+            it "prop_createPlan_small" $
+                property prop_createPlan_small
+
+        describe "Large migrations" $
+            it "prop_createPlan_large" $
+                property prop_createPlan_large
+
+        describe "Giant migrations" $
+            it "prop_createPlan_giant" $
+                property prop_createPlan_giant
+
+    parallel $ describe "Categorizing UTxO entries" $ do
+
+        it "prop_categorizeUTxOEntries" $
+            property prop_categorizeUTxOEntries
+        it "prop_categorizeUTxOEntry" $
+            property prop_categorizeUTxOEntry
+
+--------------------------------------------------------------------------------
+-- Creating migration plans
+--------------------------------------------------------------------------------
+
+data ArgsForCreatePlan = ArgsForCreatePlan
+    { mockConstraints :: MockTxConstraints
+    , mockInputs :: [(MockInputId, TokenBundle)]
+    , mockRewardWithdrawal :: Coin
+    }
+    deriving (Eq, Show)
+
+newtype Small a = Small { unSmall :: a }
+    deriving (Eq, Show)
+
+newtype Large a = Large { unLarge :: a }
+    deriving (Eq, Show)
+
+newtype Giant a = Giant { unGiant :: a }
+    deriving (Eq, Show)
+
+instance Arbitrary (Small ArgsForCreatePlan) where
+    arbitrary = Small <$> genArgsForCreatePlan
+        (0, 100) genMockInput
+
+instance Arbitrary (Large ArgsForCreatePlan) where
+    arbitrary = Large <$> genArgsForCreatePlan
+        (1_000, 1_000) genMockInput
+
+instance Arbitrary (Giant ArgsForCreatePlan) where
+    arbitrary = Giant <$> genArgsForCreatePlan
+        (100_000, 100_000) genMockInputAdaOnly
```

Comment:
With recent optimizations, I think we can relax the restriction of using ada-only inputs for the "giant" test, provided that we reduce the "giant" size to something more manageable, such as 10,000 entries. (Which should take around 3 seconds.)

Reply:
Fixed in 522a895a11384b9faebc6868997f08f836ef0fde.